gateway-provider-security-jwt/src/main/java/org/apache/knox/gateway/provider/federation/jwt/filter/SignatureVerificationCache.java

```diff
@@ -0,0 +1,141 @@
+     */
+    @SuppressWarnings("PMD.SingletonClassReturningNewInstance")
+    public static SignatureVerificationCache getInstance(final String topology, final FilterConfig config) {
+        SignatureVerificationCache instance;
+        String cacheId = topology != null ? topology : DEFAULT_CACHE_ID;
+        synchronized (instances) {
+            instance = instances.get(cacheId);
+            if (instance == null) {
+                instance = new SignatureVerificationCache(config);
+                instances.put(cacheId, instance);
+                log.initializedSignatureVerificationCache(cacheId);
+            }
+        }
+        return instance;
+    }
+
+    private Cache<String, Boolean> verifiedTokens;
+
+    private SignatureVerificationCache(final FilterConfig config) {
+        initializeVerifiedTokensCache(config);
+    }
+
+    /**
+     * Initialize the cache for token verification records.
+     *
+     * @param config The configuration of the provider employing this cache.
+     */
+    private void initializeVerifiedTokensCache(final FilterConfig config) {
+        int maxCacheSize = JWT_VERIFIED_CACHE_MAX_DEFAULT;
+
+        String configValue = config.getInitParameter(JWT_VERIFIED_CACHE_MAX);
+        if (configValue != null && !configValue.isEmpty()) {
+            try {
+                maxCacheSize = Integer.parseInt(configValue);
+            } catch (NumberFormatException e) {
+                log.invalidVerificationCacheMaxConfiguration(configValue);
+            }
+        }
+
+        verifiedTokens = Caffeine.newBuilder().maximumSize(maxCacheSize).build();
+    }
+
+    /**
+     * Determine if the specified JWT's signature has previously been successfully verified.
+     *
+     * @param jwt A serialized JWT.
+     *
+     * @return true, if the specified token has been previously verified; Otherwise, false.
+     */
+    public boolean hasSignatureBeenVerified(final String jwt) {
+        return (verifiedTokens.getIfPresent(jwt) != null);
+    }
+
+    /**
+     * Record a successful token signature verification.
+     *
+     * @param jwt A serialized JWT for which the signature has been successfully verified.
+     */
+    public void recordSignatureVerification(final String jwt) {
+        verifiedTokens.put(jwt, true);
+    }
+
+    /**
+     * Explicitly evict the signature verification record from the cache if it exists.
+     *
+     * @param jwt The serialized JWT for which the associated signature verification record should be evicted.
+     */
+    public void removeSignatureVerificationRecord(final String jwt) {
+         verifiedTokens.asMap().remove(jwt);
+    }
+
+    /**
+     * @return The size of the cache.
+     */
+    public long getSize() {
+        return verifiedTokens.estimatedSize();
+    }
+
+    /**
+     * Explicitly clean up any records which should be evicted from the cache.
+     */
+    public void cleanUp() {
```

Comment:
If I understood well, this is a `run maintenance now` (aka. evict now) method, right? It'd be better to have a name indicating this. I'd check the underlying cache implementation to understand the difference between this method and `clear()`

Reply:
cleanUp is indeed an "evict least-recently-used entries" method. Clear is what you would expect, completely clearing the contents of the cache.

Reply:
Changed it to performMaintenance()